### DISCUSSION TITLE
Scoring in `sknnr`

### DISCUSSION THREAD

**Discussion by aazuspan** [lemma-osu/sknnr#58] [Ideas]
I know there are some nuances to scoring kNN that might not apply elsewhere in `sklearn`, so we probably need to spend some more time thinking through what should be implemented in `sknnr`, what should be left up to users and demonstrated in examples, what's out of scope, etc.

@grovduck I tried to outline the topics that came up in past discussions, but please fill in any details that you think I missed!

## Per-attribute vs. aggregated scores

kNN is multi-output, so the typical `score` method will calculate mean R<sup>2</sup> of all attributes. Per-attribute scores may be more informative.

While this isn't doable with the `score` method, it is easy to apply any scoring function like `r2_score` to one or more attributes of interest. We could add a parameter to allow selecting subsets of attributes with the `score` method, but my preference would be to leave this up to users and keep `score` consistent with other estimators.

## Continuous vs. categorical attributes

kNN can predict continuous and categorical attributes (technically, although categorical attributes would only be preserved if k=1), and a single metric may not be suitable for combinations of the two. As with per-attribute scores, this may be best left up to users as defining continuous and categorical attributes when scoring would probably be more work for them than just using the metric and attributes they want. 

## Changing the `score` metric

The `score`  method calculates R<sup>2</sup> and doesn't allow you to choose other metrics. I think we settled on keeping the default scorer for consistency, but we also discussed the idea of adding a parameter to allow choosing a different metric.

## Cross-validation and grid search

We want to maintain compatibility with these. I don't think there's anything preventing that at the moment, but we should keep it in mind for any proposed changes to scoring.

## Score weighting

Quoting @grovduck in reference to [McRoberts 2012](https://www.sciencedirect.com/science/article/abs/pii/S0378112711004014):

> In section 3.3.2, third paragraph, he outlines the “T” statistic, which is roughly a user-weighted multivariate R^2 as Dave suggested.  This is what scikit-learn is doing with equal weighting.  Ron uses it as a metric for determining the value of k but we could use it more broadly.  I would think a user could either specify equal weighting or provide an array of weights equal to the number of response attributes.

To implement this, do we modify the `score` method to accept a `score_weight` parameter? Or would this be a separate scoring function that takes a weight parameter?

## Additional metrics

We also discussed that there may be additional metrics that are worth including:

- Normalized RMSE
- Bray Curtis
  - This can be calculated via [pairwise_distances](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html). I'm not sure if we might need to modify this for our use?
- Others?
